### PR TITLE
Field handling algorithm. Close #22

### LINKED ISSUE #18: Tetromino handling algorithm
After all helper functions in place we can implement class-based operations on tetromino, namely rotation and translation.
It should remember current rotation phase and check boundary conditions.
Create a class for each of tetromino types (7 classes).

You should update/access inner properties of class via constructor and methods:
- phase
- coordinates
- tetromino type (this should not change after initialization)

## Constructor
Each class should have a constructor, that will create tetromino of this type. 

Constructor should have 3 parameters:
- tetromino type (single letter)
- initial rotationPhase (string, like "horizontal")
- initial coordinates (array of (x,y)-objects) (in tests use buildTetromino)

## Rotate tetromino
There should be a method to rotate tetromino on a field

Parameters:
- field (array of arrays of booleans). If element is true - this coordinate already has a block on it. Size of outer array - is height of field (y-coordinate). Size of inner array - is width of field (x-coordinate). In tests, please construct parameter via buildField.

Function returns:
- coordinates property of class

Algorithm:
- Initiate rotation (rotateTetromino - all parameters should come from inside the class itself: typeOfTetromino, rotationPhaseOfTetromino, tetrominoCoordinates)
- check if valid position - not out-of-bounds or intersecting places blocks on field (checkThatTheFieldIsFree - move to top-level function in tetris.js, add tests for it in separate test file)
- If check result is valid - update rotation phase (incrementPhase) and coordinates
- If check result is invalid - leave tetromino coordinates and phase as they were before rotation was attempted

Unit tests
- for each type of tetromino, create it and cycle through all rotation phases and check that it returns to the original position. (7 tests)
Cover with tests following equivalence classes:
- different tetromino types
- rotation starts in different phases
- intersects with blocks on field after rotation
- goes out of bounds of field
  * too low
  * too high 
  * too to the left
  * too to the right

## Move tetromino
This function will translate tetromino position.

Parameters:
- field (same as above)
- shift - (x,y)-object. Function should do nothing if invalid move is attempted. Valid moves (shifts):
- 1 to the left (-1,0)
- 1 to the right (1,0)
- 1 down (0,1)

Function returns:
- coordinates property of class

Algorithm:
- Initiate movement - (shiftCoordinates)
- check if valid position - not out-of-bounds or intersecting places blocks on field (checkThatTheFieldIsFree)
- If check result is valid - update coordinates
- If check result is invalid - do not update coordinates

Unit tests:
One unit test for each type of tetromino for valid movement (21 tests)
- left 
- right
- down

One unit test for each type of tetromino for invalid movement:
- goes out of bounds of field
  * too low
  * too to the left
  * too to the right

### LINKED ISSUE #22: Field handling algorithm
Create and handle field 

Field contains single property 
* field - array of arrays of booleans. If element is true - this coordinate already has a block on it (it is "full"). If element is fals - this coordinate does not have a block in it (it is "empty"). Size of outer array - is height of field (y-coordinate). Size of inner array - is width of field (x-coordinate).

# Constructor
Constructor that accepts size of field and generates empty field of this size

Parameters:
- size - (x,y)-object. X will be width and y will be height of the field.

# Constructor with non-empty field
Constructor accepts pre-constructed field - array of arrays of booleans. This parameter is copied to local property 'field'. This will be useful during the testing.

# Add tetromino to field
Should mark previously "empty" field positions field property as "full" based on coordinates of tetromono.

Parameters:
- tetromino - array of (x,y)-objects.

Returns:
- field property of class.

Algorithm:
Go through all elements of tetromino, and mark corresponding elements of "field" as "full".

Add unit tests.

# Clean filled rows
We need to clear filled rows - otherwise it's not a fun game.

Parameters:
- (no parameters)

Returns:
- field property of class.

Algorithm:
go through all Y-coordinates of field property and check each row if all elements are "full" on this row (same y, different x coordinates). 
- if yes, remove this row from field, and add new row of "empty" element on top of field.
- If no, do nothing with this row.

Add unit tests:
- Check situation where there's no filled rows
- Check situation where there's 1 filled row
- Check situation where there's more than 1 filled row

### PR DESCRIPTION
**Pull request by PashaBeshchuk**

### CODE REVIEW

File: tetris.js [hunk 1]
```diff
@@ -454,12 +454,19 @@ module.exports = {
...
-		constructor(field, size){
-			this.field = field
-			this.size = size
-		}
-		generationField(size){
-			
+		constructor(field) {
+			if(Array.isArray(field)){
+				this.field = field
+			}else{
```
Comment (zbstof) [inline tetris.js]: format

File: tetris.js [hunk 2]
```diff
@@ -454,12 +454,19 @@ module.exports = {
...
+				this.filed = []
```
Comment (zbstof) [inline tetris.js]: typo

File: tetris.js [hunk 3]
```diff
@@ -452,5 +452,47 @@ module.exports = {
...
+		}
+		addTetrominoToField(coordinatesOfTetromino) {
+			for (let i = 0; i < coordinatesOfTetromino.length; i++) {
+				this.field[coordinatesOfTetromino[i].y][coordinatesOfTetromino[i].x] = true
+			}
+			return this.field
+		}
+		cleanFilledRows() {
+			let partOfTheField = []
+			for (let i = 0; i < this.field[0].length; i++) {
```
Comment (zbstof) [inline tetris.js]: you can save 
this.field[0].length
and
this.field.length
in separate field in the class. Initialize them in the constructor
Something like 
size: {x: ?, y: ?}

File: tetris.js [hunk 4]
```diff
@@ -452,5 +452,47 @@ module.exports = {
...
+			for (let i = 0; i < this.field.length; i++) {
+				let sumFullCells = 0
+				for (let j = 0; j < this.field[i].length; j++) {
+					if (this.field[i][j]) {
+						sumFullCells++
+					}
+				}
+				if (this.field.length === sumFullCells) {
+					this.field.splice(i, 1);
+					this.field.unshift(partOfTheField)
```
Comment (zbstof) [inline tetris.js]: this is not completely correct

Write tests that (remove lines, add tetrominos) - repeat several times. You will notice a bug

File: test/test-cleanFilledRows.js [hunk 5]
```diff
@@ -64,5 +67,68 @@ describe("Check situation on field", function () {
...
+        `)
+        let field = new tetris.Field(fieldSize)
+        let firstResultCheckField = field.cleanFilledRows()
+        let addTetromino = field.addTetrominoToField(coordinatesTetromino_L)
+        let secondResultCheckField = field.cleanFilledRows()
+        let addNewTetromino = field.addTetrominoToField(coordinatesTetromino_Z)
+        let thirdResultCheckField = field.cleanFilledRows()
+        let addNewTetrominoJ = field.addTetrominoToField(coordinatesTetromino_J)
+        let forthResultCheckField = field.cleanFilledRows()
+        let expectedResult = tetris.buildField(`
```
Comment (zbstof) [inline test/test-cleanFilledRows.js]: Add intermediate asserts for state of the field

Comment (PashaBeshchuk) [reply test/test-cleanFilledRows.js]: some assert.deepEqual in one "it" do not work

**Pull request merged**